File: DataStructure/merge_sort.py
def merge_sort(arr):
    if len(arr) < 2:
        return arr
    #중간점 기준으로 왼쪽, 오른쪽
    mid = len(arr) // 2
   
    low_arr = merge_sort(arr[:mid])
    high_arr = merge_sort(arr[mid:])
    return merge(low_arr, high_arr)

def merge(low_arr, high_arr):
    merged_arr = []
    l = h = 0
    while l < len(low_arr) and h < len(high_arr):
        if low_arr[l] < high_arr[h]:
            merged_arr.append(low_arr[l])
            l += 1
        else:
            merged_arr.append(high_arr[h])
            h += 1
     #  리스트 slice를 할 때 배열의 복제가 일어나므로 메모리 사용 효율은 좋지 않음
    merged_arr += low_arr[l:]
    merged_arr += high_arr[h:]
    return merged_arr

def merge2(low_arr, high_arr):
    l = h = 0
    merged_arr = []

    while(l < len(low_arr) and h < len(high_arr)):
        if low_arr[l] < high_arr[h]:
            merged_arr.append(low_arr[l])
            l +=1
        else:
            merged_arr.append(high_arr[h])
            h +=1
    while(l<len(low_arr)):
        merged_arr.append(low_arr[l])
        l +=1
    while(h<len(high_arr)):
        merged_arr.append(high_arr[h])
        h +=1
    
    return merged_arr

File: DataStructure/MergeSort.java
public class MergeSort{
    /*
    의사코드
    mergeSort(arr[], int l, int r){
        if(l < r){
            int mid = (l+r)/2;
            mergeSort(arr, l, mid);
            mergeSort(arr, mid+l, r);
            merge(arr, l, mid, r);
        }
    }

    merge(arr[], int l, int mid, int r){
        정렬된 두 배열 (arr[l..mid]), (arr[mid+l..r])를 합병
        정렬된 하나의 배열 (arr[l..r])을 도출
    }
    */

    public static void mergeSort(int[] arr, int l, int r){
        if(l < r){
            int mid = (l+r)/2;
            mergeSort(arr, l, mid);
            mergeSort(arr, mid+1, r);
            merge(arr, l, mid, r);
        }
    }

    public static void merge(int[] arr, int l, int mid, int r){
        int i, k;
        i = k = l;
        int j = mid +1;
        int[] temp = new int[arr.length];
        while(i <= mid && j <= r){
            if(arr[i] < arr[j]){
                temp[k++] = arr[i++];
            }else{
                temp[k++] = arr[j++];
            }
        }

        while(i <= mid){
            temp[k++] = arr[i++];
        }
        while(j <= r){
            temp[k++] = arr[j++];
        }

        for(int index=l; index<k; index++){
            arr[index] = temp[index];
        }
    }

    public static void printArray(int[] arr){
        for(int i = 0; i < arr.length; i++){
            System.out.print(arr[i] + " ";)
        }
        System.out.println();
    }

    public static void main(String[] args){
        int arr[] = {6,3,3,7,1,21,3,7,2,25};
        mergeSort(arr, 0, arr.length - 1);
        printArray(arr);
    }
}